cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -784,43 +865,103 @@ public void receivedPollingInformation(@Nonnull StreamSize pollingInfo) {
           argsBuilder.put(ProgramOptionConstants.PAST_RUN_DATA_SIZE, Long.toString(lastRunSize));
         }
 
+        try {
+          scheduleStore.updateBaseRun(programId, programType, streamSizeSchedule.getName(),
+                                      pollingInfo.getSize(), pollingInfo.getTimestamp());
+        } catch (Throwable t) {
+          LOG.error("Error when persisting new base information for schedule {} in store. Program will not be executed",
+                    streamSizeSchedule.getName(), t);
+          return;
+        }
+
         currentSchedule = streamSizeSchedule;
         basePollSize = pollingInfo.getSize();
         basePollTs = pollingInfo.getTimestamp();
       }
 
       ScheduleTaskRunner taskRunner = new ScheduleTaskRunner(store, programRuntimeService, preferencesStore);
+      long previousLastRunSize = lastRunSize;
+      long previousLastRunTs = lastRunTs;
       try {
+        try {
+          scheduleStore.updateLastRun(programId, programType, streamSizeSchedule.getName(),
+                                      pollingInfo.getSize(), pollingInfo.getTimestamp());
+          lastRunSize = pollingInfo.getSize();
+          lastRunTs = pollingInfo.getTimestamp();
+        } catch (Throwable t) {
+          LOG.error("Error when persisting last run information for schedule {} in store",
+                    streamSizeSchedule.getName(), t);
+          return;
+        }
+        // The taskRunner.run() call will wait for the program to be run entirely, and can't be wrapped in
```

Comment:
``TaskExecutionException`` is thrown when the program can't be found in the store, or if it is already running and the option to set concurrent runs is not on, or if waiting for the program to finish has been interrupted.